### DISCUSSION TITLE
Improvements to Dynamic Components, aka <component :is />

### DISCUSSION THREAD

**Discussion by BlackYuzia** [vuejs/rfcs#609] [General]
Hello there.

### Description

How you could know vue has [dynamic-components](https://vuejs.org/guide/essentials/component-basics.html#dynamic-components)
That will allow to use change component fast and simple (without a lot of `if-else`, `v-else-if` and `v-else`). Hello `v-switch` & `v-switch-case`? But that a bit off-topic now.

### Problem

So, we could make something like this:
```vue
<template comments="remove">
    <component :is="getComponent()"></component>
</template>


<script lang="ts" setup comments="remove">
function getComponent() {
    // VHelloWorld1 has [title, description] props, title - required
    // VHelloWorld2 has [active], active - required
    return Math.random() > 0.5 ? defineAsyncComponent(() => import('~/components/VHelloWorld1.vue')) : defineAsyncComponent(() => import('~/components/VHelloWorld2.vue'))
}
</script>
```

That code will return to us `VHelloWorld1` or `VHelloWorld2` component.
But, we have bad news. We don't get any warns about props that not exist in our component:
```vue
<component :is="getComponent()"></component>
```

VSCode **do not** say to us "hello, you forget `title`, `description` and `active` props,".
That's a huge trouble when you add / change any props and then found what you forgot to add it to the `<component :is="...">`

### Solutions

I see only one way to fix that properly, maybe **you** know any other ways.

-  Vue should "say" to **vscode** (IDE) what current type has `<component>` (merge all possible types from `:is="getComponent()"`) and return merged props (and/or any else too)

This will grant what we don't miss any prop(s) on changing components or components list

### Potential difficulties

If we look our component with vscode we saw what `component` do not has type as well.
![image](https://github.com/vuejs/rfcs/assets/25010528/858df56c-93af-430f-9472-0ba2734de2e2)
and if we click on it with `CTRL + LMB` nothing happen. That because `<component>` is **DYNAMIC COMPONENT!, WOW**.
![image](https://github.com/vuejs/rfcs/assets/25010528/d8e733a2-d8b8-4523-9822-cb42fe70316e)

That mean what he does not exist in our components.

But, this will work with other components.
![image](https://github.com/vuejs/rfcs/assets/25010528/5c1637bc-79e6-4cee-b777-0f264e428a56)
![image](https://github.com/vuejs/rfcs/assets/25010528/ef19d62e-2d0d-4630-9d15-d088b3afd2ce)

### Conclusions

We have dynamic components.
We can't set or get the type of these components in any case, right?
We want types when it is possible.
We need to make it possible if no types there.

**Thanks for attention.**

**Comment by sadeghbarati** [comment]
Related thread in Volar discord

https://discord.com/channels/793943652350427136/1099776004286844938/1100032028927340705

I'll test with dynamic import cause I don't remember what should I expect from Vue Language Tool when using `defineAsyncComponent` and `<component :is="...">`


If you can confirm issue maybe it's better to report the issue in Vue Language Tools repository
